### PR TITLE
ci: integrate goreleaser

### PR DESCRIPTION
**Pull request by kaustubhkapatral**
Closes https://github.com/cosmos/iavl/issues/981

Note:- The `RELEASE_NOTES.md` file present in the PR has placeholder text which was taken from the release page of [v1.1.0](https://github.com/cosmos/iavl/releases/tag/v1.1.0). Please update that file before pushing a tag along with the changelog..

## Summary by CodeRabbit

- **New Features**
  - Introduced an automated release workflow to streamline project releases upon tagging.
  - Added configuration for GoReleaser to manage versioning and changelog generation.
  
- **Documentation**
  - Updated release notes to highlight API-breaking changes and their implications for upcoming SDK versions. 

- **Chores**
  - Created a new workflow file to enhance deployment efficiency.

### CODE REVIEW

**Comment by kaustubhkapatral** [comment]
@tac0turtle  Would appreciate a review on this

File: RELEASE_NOTES.md
Comment (julienrbrt) [inline RELEASE_NOTES.md:1]: Hey, we don't tag from main, so this file shouldn't be there. That's something we should create in the release process when tagging.

**Review by julienrbrt — APPROVED** [review]
lgtm!, one nit

**Comment by kaustubhkapatral** [comment]
@julienrbrt Got it. I've removed the `RELEASE_NOTES.md` file from the PR and merged the latest master commit to my fork and PR

**Review by julienrbrt — APPROVED** [review]

**Pull request merged**